sled-agent/src/bootstrap/early_networking.rs

```diff
@@ -38,59 +47,238 @@ pub enum EarlyNetworkSetupError {
     Dendrite(String),
 
     #[error("Error during DNS lookup: {0}")]
-    DnsResolver(#[from] internal_dns::resolver::ResolveError),
+    DnsResolver(#[from] ResolveError),
+}
+
+enum LookupSwitchZoneAddrsResult {
+    // We found every switch zone reported by internal DNS.
+    TotalSuccess(HashMap<SwitchLocation, Ipv6Addr>),
+    // We found some (but not all) switch zones reported by internal DNS.
+    PartialSuccess(HashMap<SwitchLocation, Ipv6Addr>),
 }
 
 /// Code for configuring the necessary network bits to bring up the control
 /// plane
-pub struct EarlyNetworkSetup {
-    log: Logger,
+pub struct EarlyNetworkSetup<'a> {
+    log: &'a Logger,
 }
 
-impl EarlyNetworkSetup {
-    pub fn new(log: &Logger) -> Self {
-        EarlyNetworkSetup { log: log.clone() }
+impl<'a> EarlyNetworkSetup<'a> {
+    pub fn new(log: &'a Logger) -> Self {
+        EarlyNetworkSetup { log }
+    }
+
+    pub async fn lookup_boundary_switch_addrs(
+        &self,
+        resolver: &DnsResolver,
+        config: &RackNetworkConfig,
+    ) -> HashSet<Ipv6Addr> {
+        let switch_zone_addrs = self.lookup_switch_zone_addrs(resolver).await;
+        let mut boundary_switch_addrs = HashSet::new();
+
+        for uplink_config in &config.uplinks {
+            if let Some(&zone_addr) =
+                switch_zone_addrs.get(&uplink_config.switch)
+            {
+                boundary_switch_addrs.insert(zone_addr);
+            } else {
+                // `lookup_boundary_switch_addrs` is best effort: if one of the
+                // switches is offline (or we can't find it for any number of
+                // reasons), we may not have the address for this
+                // `uplink_config`. Just log a warning and move on.
+                warn!(
+                    self.log,
+                    "No switch zone address found for {:?}",
+                    uplink_config.switch,
+                );
+            }
+        }
+
+        // TODO-correctness: Could `boundary_switch_addrs` be _empty_ here?
```

Comment:
At this point, we also will not be able to sync time and bring up the rest of the control plane. I think it's worth an explicit error being passed upwards or an error log and a panic. Ideally we'd be able to show this in the UI, but for now I think the easiest thing to do is log an error and panic.

Reply:
This isn't only used during RSS - it's also used any time any sled-agent starts an external DNS / Nexus / NTP service. So for example, we could have this very unlucky sequence:

1. Two switches are present at RSS
2. Only one switch has an uplink
3. We start RSS. NTP services start, and they all call this function - both switches are found, and we filter down to the one switch that has an uplink in this function
4. NTP sync succeeds
5. The one switch zone one connected to the switch with an uplink goes down
6. We try to start nexus / external DNS: we get to this function, only find one switch, and then filter it away because it doesn't have an uplink
7. We return an empty `boundary_switch_addrs`, so nexus and external DNS do not set up NATs (which they _can't_ do anyway, because the one switch with an uplink is down)

I think I have convinced myself the actual check should not be "we found at least one switch", but "we found at least one switch _with an uplink_", and we should block until we do? I'll make that change, and if you disagree I can back it out.

Reply:
This change looks great! Thank you so much!